homeassistant/components/sonos/number.py

```diff
@@ -142,3 +173,408 @@ def native_value(self) -> float:
+
+        if self._value != vol:
+            self._value = vol
+            self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Bind signals, subscribe to group channels, perform initial populate/refresh."""
+        await super().async_added_to_hass()
+
+        self._coord_uid = (self.speaker.coordinator or self.speaker).uid
+        self._group_uid = self._current_group_uid()
+
+        # Rebind/refresh on any speaker activity
+        self._unsubscribe_activity = async_dispatcher_connect(
+            self.hass, SONOS_SPEAKER_ACTIVITY, self._on_any_activity
+        )
+        self.async_on_remove(self._unsubscribe_activity)
+
+        # Coordinator state updates
+        self._unsubscribe_coord = async_dispatcher_connect(
+            self.hass,
+            f"{SONOS_STATE_UPDATED}-{self._coord_uid}",
+            self._on_coord_state_updated,
+        )
+        self.async_on_remove(self._unsubscribe_coord)
+
+        # This member's own updates
+        self._unsubscribe_member = async_dispatcher_connect(
+            self.hass,
+            f"{SONOS_STATE_UPDATED}-{self.speaker.uid}",
+            self._on_member_state_updated,
+        )
+        self.async_on_remove(self._unsubscribe_member)
+
+        # Subscribe to current group fan-out (if any)
+        self._subscribe_group_fanout(self._group_uid)
+
+        # Coordinator listens for refresh requests for its group (if applicable)
+        self._subscribe_group_requests_if_coord(self._group_uid)
+
+        # Kick an initial rebind + populate, then a delayed settle read
+        self._rebind_for_topology_change()
+        await self._async_initial_populate()
+        self._schedule_delayed_refresh(GV_REFRESH_DELAY)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Cleanup on removal (dispatcher unsubs handled elsewhere; cancel pending timer)."""
+        await super().async_will_remove_from_hass()
+
+        # Let async_on_remove(...) handle dispatcher unsubs; just cancel timer
+        if self._delay_unsubscribe is not None:
+            self._delay_unsubscribe()
+            self._delay_unsubscribe = None
+
+    @callback
+    def _on_group_volume_request(self, *_: object) -> None:
+        """Coordinator-only: handle a member’s request to refresh group volume."""
+        if not (self._is_grouped() and self._is_coordinator()):
+            return
+        self._schedule_delayed_refresh(GV_REFRESH_DELAY)
+
+    @callback
+    def _on_coord_state_updated(self, *_: object) -> None:
+        """Coordinator state changed; re-evaluate bindings and schedule refresh if needed."""
+        self._rebind_for_topology_change()
+
+    @callback
+    def _on_member_state_updated(self, *_: object) -> None:
+        """Member state changed; rebind and trigger the appropriate refresh path."""
+        self._rebind_for_topology_change()
+
+    @callback
+    def _on_group_volume_fanned(self, payload: tuple[str, int]) -> None:
```

Comment:
I like this, great way to fan out the volume and not having to refresh it on each entity of the group